{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: limma\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Loading required package: parallel\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:limma’:\n",
      "\n",
      "    plotMA\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which, which.max, which.min\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Loading required package: DelayedArray\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "Loading required package: BiocParallel\n",
      "\n",
      "\n",
      "Attaching package: ‘DelayedArray’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    aperm, apply, rowsum\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(edgeR)\n",
    "library(magrittr)\n",
    "library(ggplot2)\n",
    "library(DESeq2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Automatise"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# On cell-state level"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## For LV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "x=read.csv(\"/fast/AG_Huebner/huebner3/ANALYSES/20200706_el_COVID_Biopsy/data/COVID_ALL_CD4CD8NK_PSEUDOBULK_LV.csv\")\n",
    "genes_tofilter=read.csv(\"/fast/AG_Huebner/huebner3/ANALYSES/20200706_el_COVID_Biopsy/data/COVID_ALL_CD4CD8NK_PSEUDOBULK_FILTERING_LV.csv\")\n",
    "\n",
    "# Only needed for the column cell_state\n",
    "CELLTYPE_STATE <- read.csv(\"/fast/AG_Huebner/huebner3/ANALYSES/20200706_el_COVID_Biopsy/data/COVID_CD4CD8NK_TRANSLATION_TABLE.csv\")\n",
    "colnames(CELLTYPE_STATE) <- c(\"cell_states\", 'cell_type')\n",
    "\n",
    "# Removes patients with only few (<10) nuclei numbers (as they might bias the analysis)\n",
    "CELLTYPE_FILTER <- read.csv(\"/fast/AG_Huebner/huebner3/ANALYSES/20200706_el_COVID_Biopsy/data/COVID_ABSOLUTE_CELLTYPE_NUMBER_LV.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "PATIENTS_TOKEEP <- CELLTYPE_FILTER[CELLTYPE_FILTER$celltype_classified==\"Lymphoid\",2:43]>=30"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "PATIENTS_TOKEEP <- names(which(PATIENTS_TOKEEP[1,]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "rownames(x) <- x$X\n",
    "x <- x[,-1]\n",
    "l <- strsplit(colnames(x), \"__\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta.data <- t(as.data.frame(l))\n",
    "rownames(meta.data) <- colnames(x)\n",
    "\n",
    "meta.data <- as.data.frame(meta.data)\n",
    "colnames(meta.data)  <-  c(\"Condition_custom\", \"cell_state\", 'Patient')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>Myocarditis</li><li>Pims</li><li>Post.Covid.19</li><li>Post.Vaccination</li></ol>\n",
       "\n",
       "<details>\n",
       "\t<summary style=display:list-item;cursor:pointer>\n",
       "\t\t<strong>Levels</strong>:\n",
       "\t</summary>\n",
       "\t<style>\n",
       "\t.list-inline {list-style: none; margin:0; padding: 0}\n",
       "\t.list-inline>li {display: inline-block}\n",
       "\t.list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "\t</style>\n",
       "\t<ol class=list-inline><li>'Control'</li><li>'Myocarditis'</li><li>'Pims'</li><li>'Post.Covid.19'</li><li>'Post.Vaccination'</li></ol>\n",
       "</details>"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item Myocarditis\n",
       "\\item Pims\n",
       "\\item Post.Covid.19\n",
       "\\item Post.Vaccination\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\emph{Levels}: \\begin{enumerate*}\n",
       "\\item 'Control'\n",
       "\\item 'Myocarditis'\n",
       "\\item 'Pims'\n",
       "\\item 'Post.Covid.19'\n",
       "\\item 'Post.Vaccination'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. Myocarditis\n",
       "2. Pims\n",
       "3. Post.Covid.19\n",
       "4. Post.Vaccination\n",
       "\n",
       "\n",
       "\n",
       "**Levels**: 1. 'Control'\n",
       "2. 'Myocarditis'\n",
       "3. 'Pims'\n",
       "4. 'Post.Covid.19'\n",
       "5. 'Post.Vaccination'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] Myocarditis      Pims             Post.Covid.19    Post.Vaccination\n",
       "Levels: Control Myocarditis Pims Post.Covid.19 Post.Vaccination"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "GENOTYPES <- unique(meta.data[,\"Condition_custom\"][meta.data[,\"Condition_custom\"] != \"Control\"])\n",
    "GENOTYPES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "CELL_LEVEL <- \"CELLSTATES\"\n",
    "REGION <- \"LV\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "i <- 1\n",
    "\n",
    "for(GENOTYPE in as.character(GENOTYPES)){\n",
    "    message(\"\\n###START: \", GENOTYPE,\" #####\\n\")\n",
    "    \n",
    "    for(CELL_STATE in as.character(unique(meta.data$cell_state))){\n",
    "    #for(CELL_STATE in c(\"EC7\")){\n",
    "        \n",
    "        CONTROL_COLUMN <- paste(\"control_\",CELL_STATE, sep=\"\")\n",
    "        GENOTYPE_COLUMN <- paste(GENOTYPE,\"_\", CELL_STATE, sep=\"\")\n",
    "        CELLTYPE <- as.character(CELLTYPE_STATE[which(CELLTYPE_STATE$cell_states==CELL_STATE),\"cell_type\"])\n",
    "        \n",
    "        \n",
    "        x_sub <- x[,which(meta.data[,\"Condition_custom\"] %in% c(\"Control\", GENOTYPE))]\n",
    "        meta.data_sub <- meta.data[which(meta.data[,\"Condition_custom\"] %in% c(\"Control\", GENOTYPE)),]\n",
    "        \n",
    "        x_sub <- x_sub[,which(meta.data_sub$cell_state %in% c(CELL_STATE))]\n",
    "        meta.data_sub <- meta.data_sub[which(meta.data_sub$cell_state %in% c(CELL_STATE)),]\n",
    "        \n",
    "        ##### remove first line (cell_type column)\n",
    "        #CELLTYPE_FILTER_SUB <- CELLTYPE_FILTER\n",
    "        #PATIENTS_TOKEEP <- CELLTYPE_FILTER_SUB[which(CELLTYPE_FILTER_SUB$cell_state==CELL_STATE),-1]\n",
    "        #PATIENTS_TOKEEP <- colnames(PATIENTS_TOKEEP)[which(as.logical(PATIENTS_TOKEEP))]\n",
    "        #####\n",
    "        \n",
    "        # Subset only for patients which fulfill condition\n",
    "        x_sub <- x_sub[,which(meta.data_sub$Patient %in% PATIENTS_TOKEEP)]\n",
    "        meta.data_sub <- meta.data_sub[which(meta.data_sub$Patient %in% PATIENTS_TOKEEP),]\n",
    "        \n",
    "        meta.data_sub[,\"Condition_custom\"] <- factor(as.character(meta.data_sub[,\"Condition_custom\"]))\n",
    "        #meta.data_sub[,\"Condition_custom\"] = droplevels(meta.data_sub[,\"Condition_custom\"]) ## Drop factor levels\n",
    "   \n",
    "        \n",
    "        # The test is only run, if 2 conditions have at least 3 patients\n",
    "        if (length(unique(meta.data_sub[,\"Condition_custom\"]))==2 & \n",
    "            all(table(as.character(meta.data_sub[,\"Condition_custom\"]))>=3))\n",
    "           {\n",
    "             # Prepare DGEList object\n",
    "            meta.data_sub[,\"Condition_custom\"] <- as.factor(as.character(meta.data_sub[,\"Condition_custom\"]))\n",
    "            dge <- DGEList(counts=x_sub, group=meta.data_sub[,\"Condition_custom\"])\n",
    "            \n",
    "            # Filter genes based on expression, the qlf Object will be filtered to exclude them from FDR calculation\n",
    "            genes_tofilter_sub_GENOTYPE <- genes_tofilter[,which(gsub(\"^(.*)__.*__.*$\", \"\\\\1\", colnames(genes_tofilter)) %in% GENOTYPE &\n",
    "                                                                 gsub(\"^.*__(.*)__.*$\", \"\\\\1\", colnames(genes_tofilter)) %in% CELL_STATE &\n",
    "                                                                 gsub(\"^.*__.*__(.*)$\", \"\\\\1\", colnames(genes_tofilter)) %in% PATIENTS_TOKEEP)]\n",
    "\n",
    "            genes_tofilter_sub_CONTROL <- genes_tofilter[,which(gsub(\"^(.*)__.*__.*$\", \"\\\\1\", colnames(genes_tofilter)) %in% 'Control' &\n",
    "                                                                gsub(\"^.*__(.*)__.*$\", \"\\\\1\", colnames(genes_tofilter)) %in% CELL_STATE &\n",
    "                                                                 gsub(\"^.*__.*__(.*)$\", \"\\\\1\", colnames(genes_tofilter)) %in% PATIENTS_TOKEEP)]\n",
    "            \n",
    "            genes_tofilter_sub <- data.frame(\"X\"=genes_tofilter[,\"X\"],\n",
    "                                             CONTROL_COLUMN=apply(genes_tofilter_sub_CONTROL, 1, mean),\n",
    "                                             GENOTYPE_COLUMN=apply(genes_tofilter_sub_GENOTYPE, 1, mean))\n",
    "            colnames(genes_tofilter_sub) <- c(\"X\", CONTROL_COLUMN, GENOTYPE_COLUMN)\n",
    "            \n",
    "            keep <- which(genes_tofilter_sub[,CONTROL_COLUMN] > 0.0125 |\n",
    "                          genes_tofilter_sub[,GENOTYPE_COLUMN] > 0.0125)\n",
    "            \n",
    "            #keep <- genes_tofilter[,CONTROL_COLUMN]>0.0125 | ##### change & to | ##### \n",
    "            #        genes_tofilter[,GENOTYPE_COLUMN]>0.0125\n",
    "            #dge <- dge[keep, , keep.lib.sizes=FALSE]\n",
    "    \n",
    "            # PP, model matrix, https://www.nature.com/articles/nmeth.4612 (edgeRQLFDetRate)\n",
    "            #dge <- calcNormFactors(dge)\n",
    "            #cdr <- scale(colMeans(x_sub > 0))\n",
    "            #design <- model.matrix(~ cdr + meta.data_sub[,\"Condition_custom\"])\n",
    "            design <- model.matrix(~ meta.data_sub[,\"Condition_custom\"])\n",
    "    \n",
    "            # \n",
    "            dge <- estimateDisp(dge, design = design)\n",
    "            fit <- glmQLFit(dge, design = design)\n",
    "            qlf <- glmQLFTest(fit)\n",
    "    \n",
    "            # For all\n",
    "            tt <- topTags(qlf, n = Inf)\n",
    "            \n",
    "            # Only for \"expressed/detected\" genes\n",
    "            tt_filtered <- topTags(qlf[keep,], n = Inf)\n",
    "            \n",
    "            tt$table[,\"Gene\"] <- as.character(rownames(tt$table))\n",
    "            tt_filtered$table[,\"Gene\"] <- as.character(rownames(tt_filtered$table))\n",
    "            tt_filtered$table[,\"low_expression\"] <- \"F\" #####  Low-expression column ##### \n",
    "            tt_filtered$table[,\"FDR_plot\"] <- -log10(tt_filtered$table$FDR) #####  For plotting Volcano, here we use FDR, not pValue  ##### \n",
    "            \n",
    "            \n",
    "            tt_merged <- merge(tt$table[,c(\"Gene\", \"logFC\", \"logCPM\", 'F', 'PValue', 'FDR')], \n",
    "                               tt_filtered$table[,c(\"Gene\", \"FDR\", \"low_expression\", \"FDR_plot\")], \n",
    "                               by=\"Gene\", all=T)   \n",
    "            \n",
    "            # Replace NA FDRs with NA                   #####  ORDER CHANGED #####  \n",
    "            tt_merged[which(is.na(tt_merged$FDR)),\"FDR\"] <- 1 \n",
    "            \n",
    "            EXPRESSION_MEAN <- genes_tofilter_sub[,c(\"X\", CONTROL_COLUMN, GENOTYPE_COLUMN)]\n",
    "            colnames(EXPRESSION_MEAN) <- c(\"Gene\", \"mean_exp_control\", 'mean_exp_genotype')\n",
    "            tt_merged <- merge(tt_merged, EXPRESSION_MEAN, by=\"Gene\")\n",
    "    \n",
    "            \n",
    "            tt_merged[,\"Region\"] <- REGION\n",
    "            tt_merged[,\"annotation_level\"] <- CELL_LEVEL\n",
    "            tt_merged[,\"cell_state\"] <- CELL_STATE\n",
    "            tt_merged[,\"cell_type\"] <- CELLTYPE\n",
    "            \n",
    "            \n",
    "            tt_merged[,\"comparison\"] <- paste(\"control_\", GENOTYPE, sep=\"\")\n",
    "            \n",
    "            ###### Suggestion to add those columns\n",
    "            tt_merged[,\"Observations_genotype\"] <- as.numeric(table(meta.data_sub[,\"Condition_custom\"])[GENOTYPE])\n",
    "            tt_merged[,\"Observations_reference\"] <- as.numeric(table(meta.data_sub[,\"Condition_custom\"])['Control'])\n",
    "            \n",
    "            # low expression add T\n",
    "            tt_merged[which(is.na(tt_merged$low_expression)),\"low_expression\"] <- \"T\"\n",
    "            \n",
    "            if(i==1){\n",
    "                 final_df <- tt_merged   \n",
    "            } else {\n",
    "                final_df <- rbind(final_df, tt_merged)\n",
    "            }\n",
    "            \n",
    "            i <- i + 1\n",
    "            message(\"\\n###FINISHED: \", CELL_STATE,\" #####\\n\")\n",
    "        } else {\n",
    "            message(\"\\n###Skip: \", CELL_STATE,\" #####\\n\")\n",
    "        }\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames(final_df) <- gsub(\"FDR.x\", \"FDR_all\", colnames(final_df))\n",
    "colnames(final_df) <- gsub(\"FDR.y\", \"FDR_onlyhigh\", colnames(final_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "#write.csv(final_df, \"/fast/AG_Huebner/huebner3/ANALYSES/20200706_el_COVID_Biopsy/data/COVID_CD4CD8NK_ALLGENOTYPES_EDGER_LV_V4_NOTTM_CUTOFF30.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
